src/sage/knots/link.py

```diff
@@ -2096,18 +2105,34 @@ def khovanov_polynomial(self, var1='q', var2='t', base_ring=ZZ):
 
         .. SEEALSO:: :meth:`khovanov_homology`
         """
-        L = LaurentPolynomialRing(base_ring, [var1, var2])
-        ch = base_ring.characteristic()
+        if base_ring:
+            ring = base_ring
+            from sage.misc.superseded import deprecation
+            deprecation(40149, "base_ring is deprecated, use argument ring instead.")
+
+        ch = ring.characteristic()
+        integral = False
+        if ch == 0 and not ring.is_field():
+            integral = True
+            L = LaurentPolynomialRing(ZZ, [var1, var2, torsion])
+        else:
+            L = LaurentPolynomialRing(ZZ, [var1, var2])
         coeff = {}
-        kh = self.khovanov_homology()
+        kh = self.khovanov_homology(ring=ring)
         from sage.rings.infinity import infinity
         for h in kh:
             for d in kh[h]:
                 H = kh[h][d]
-                gens = [g for g in H.gens() if g.order() == infinity or ch.divides(g.order())]
-                l = len(gens)
-                if l:
-                    coeff[(h, d)] = l
+                gens = {g: g.order() for g in H.gens()}
+                if integral:
+                    for tor in gens.values():
+                        l = len([g for g in gens if gens[g] == tor])
+                        if tor is infinity:
+                            coeff[(h, d, 0)] = l
+                        else:
+                            coeff[(h, d, tor)] = l
```

Comment:
```suggestion
                    tor_count = {}
                    for g, tor in gens.items():
                        if tor in tor_count:
                            tor_count[tor] += 1
                        else:
                            tor_count[tor] = 1
                    for tor, ell in tor_count.items():
                        if tor is infinity:
                            coeff[(h, d, 0)] = ell
                        else:
                            coeff[(h, d, tor)] = ell
```
I think this is faster since you won't repeatedly set the coefficient. (Although such benefits only are really gained if `gens` is not "small".)